docs/proposals/rescheduling.md

```diff
@@ -0,0 +1,496 @@
+
+## Design proposal
+
+The key mechanisms and components of the proposed design are priority, preemption,
+disruption budgets, the `/evict` subresource, and the rescheduler.
+
+### Priority
+
+#### Motivation
+
+[Quality of Service levels](resource-qos.md) allow Kubernetes to safely overcommit nodes.
+Just as it is useful to overcommit nodes to increase node-level utilization, it is useful
+to overcommit clusters to increase cluster-level utilization.  Scheduling priority (which
+we abbreviate as *priority*, in combination with disruption budgets (described in the
+next section), allows Kubernetes to safely overcommit clusters much as QoS levels allow
+it to safely overcommit nodes.
+
+Today, cluster sharing among users, workload types, etc. is regulated via the
+[quota](../admin/resourcequota/README.md) mechanism. When allocating quota, a cluster
+administrator has two choices: (1) the sum of the quotas is less than or equal to the
+capacity of the cluster, or (2) the sum of the quotas is greater than the capacity of the
+cluster (that is, the cluster is overcommitted).  (1) is likely to lead to cluster
+under-utilization, while (2) is unsafe in the sense that someone's pods may go pending
+indefinitely even though they are still within their quota. Priority makes cluster
+overcommitment (i.e. case (2)) safe by allowing users and/or administrators to identify
+which pods should be allowed to run, and which should go pending, when demand for cluster
+resources exceeds supply to due to cluster overcommitment.
+
+Priority is also useful in some special-cause scenarios, such as ensuring that system
+DaemonSets can always schedule and reschedule onto every node where they want to run
+(assuming they are given the highest priority), e.g. see #21767.
+
+#### Specifying priorities
+
+We propose to add a required `Priority` field to `PodSpec`. Its value type is string, and
+the cluster administrator defines a total ordering on these strings (for example
+`Critical`, `Normal`, `Preemptible`). We choose string instead of integer so that it is
+easy for an administrator to add new priority levels in between existing levels, to
+encourage thinking about priority in terms of user intent and avoid magic numbers, and to
+make the internal implementation more flexible.
+
+When a scheduler is scheduling a new pod P and cannot find any node that meets all of P's
+scheduling predicates, it is allowed to evict ("preempt") one or more pods that are at
+the same or lower priority than P (subject to disruption budgets, see next section) from
+a node in order to make room for P, i.e. in order to make the scheduling predicates
+satisfied for P on that node.  (Note that when we add cluster-level resources (#19080),
+it might be necessary to preempt from multiple nodes, but that scenario is outside the
+scope of this document.)  The preempted pod(s) may or may not be able to reschedule. The
```

Comment:
I agree. But it still racy as long as schedulers work at the same time. My idea is just pause schedulers for a short time. We use pessimistic conflic control with deadline: reschedling algorithm will have a deadline, rescheduler must ensure that rescheduling will finished before the deadline. If it just couldn't, rescheduling algorithm will return a suboptimal(but global) pod placement, if fortunately, it can finished before deadline, we get a optimal solution. Such a implementation could ensure all schedulers will not paused for a long time. Even if we could not get the best cluster layout after one rescheduling, we could get a better and better one.

Pseudo code is like:

```
// take as input all of the nodes, all of the pods and their constraints, and current pod placements as input. 
// return a new global mapping of pods to nodes, if return before deadline, we get a optimal solution, otherwise,
// we get a suboptimal solution 
func Rescheduling(pods []api.Pod, nodes []api.Node, deadline time.Duraion)(result []Binding) {
	var result []Binding
	select {
		case <-time.After(deadline):
			// calculate can not finish before the deadline, return with a suboptimal global pod placement
			return
		default: 
			// calculate the optimal global pod placement
			return	
	}
}
```

So, if we design like this, we can solve the race problem, and we could harm user experience as less as possible.

May be premature, but we could discuss it.